### PR TITLE
Handle authenticated users when HYPR is used as a second factor

### PR DESCRIPTION
**Pull request by Yasasr1**
This pr add the implementations required to get the username of the authenticated user if there is an already authenticated user.

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java [hunk 1]
```diff
@@ -223,7 +230,23 @@ private void redirectHYPRLoginPage(HttpServletResponse response, AuthenticationC
...
                                                    AuthenticationContext context) throws AuthenticationFailedException {
 
-        String username = request.getParameter(HYPR.USERNAME);
+        String username = null;
+        if (context.getSequenceConfig() != null) {
+            Map<Integer, StepConfig> stepConfigMap = context.getSequenceConfig().getStepMap();
+            // loop through the authentication steps and find the authenticated user from the subject identifier step.
+            if (stepConfigMap != null) {
+                for (StepConfig stepConfig : stepConfigMap.values()) {
+                    if (stepConfig.getAuthenticatedUser() != null && stepConfig.isSubjectIdentifierStep()) {
```
Comment (wenodpathirana) [inline components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java:239]: I think better to change the order here
first check whether this step is the subject identifier step and then check the authenticated user is not null.

File: components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java [hunk 2]
```diff
@@ -169,7 +170,13 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
                                                  AuthenticationContext context) throws AuthenticationFailedException {
 
         try {
-            redirectHYPRLoginPage(response, context, null);
+            if (context.getLastAuthenticatedUser() != null) {
```
Comment (wenodpathirana) [inline components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java:173]: should we check this in `process` method?

File: components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java (hunk 1 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java]: addressed with 0bfcecd

File: components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java (hunk 2 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java]: it is better to check in the process method. moved with 0bfcecd

File: components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java [hunk 3]
```diff
@@ -223,7 +224,23 @@ private void redirectHYPRLoginPage(HttpServletResponse response, AuthenticationC
     private void initiateHYPRAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
                                                    AuthenticationContext context) throws AuthenticationFailedException {
 
-        String username = request.getParameter(HYPR.USERNAME);
+        String username = null;
+        if (context.getSequenceConfig() != null) {
+            Map<Integer, StepConfig> stepConfigMap = context.getSequenceConfig().getStepMap();
+            // loop through the authentication steps and find the authenticated user from the subject identifier step.
+            if (stepConfigMap != null) {
```
Comment (Yasasr1) [inline components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java:231]: use isNotEmpty

**Pull request merged**